### ISSUE TITLE
How to pass exogenous variables in the function bvar

### ISSUE THREAD

**Issue by arnab13061989** [nk027/bvar#70]
Hi,
I have few exogenous variables in my dataset which are to be controlled for . Can the same be handled using bvar() as I haven't came across any arguments in the documents.

Regards
Arnab

**Comment by nk027** [comment]
Hey,
You can do this manually by using the [Frisch-Waugh-Lovell theorem](https://en.wikipedia.org/wiki/Frisch%E2%80%93Waugh%E2%80%93Lovell_theorem). Here's some code for a univariate model (`variables` are the ones not to factor out):

```r
Q_fwl <- qr.Q(qr(X[, -variables, drop = FALSE]))
y <- y - Q_fwl %*% crossprod(Q_fwl, y)
X <- X[, variables, drop = FALSE] - Q_fwl %*%
crossprod(Q_fwl, X[, variables, drop = FALSE])
```

**Comment by arnab13061989** [comment]
Hi Nikolas,

I have few queries on the same . If I’m not wrong applying FRW would also mean transformation of Y(endogenous variables) and the variance covariance matrix of residuals . The posterior draws will be conditional on these transformed variables . Moreover, how do I interpret IRF , FEVD ? if these are done on the transformed variables .

If the above concern can be addressed , I was thinking for doing the multiplier analysis for my exogenous variables.

Regards
Arnab

**Comment by nk027** [comment]
Hey Arnab,
Yes, for a VAR you basically just need to transform Y and feed it into `bvar`. The posterior draws will effectively be draws from a VARX (albeit a bit hacky, since you'd just be using the posterior mean for FWL), so the coefficients are already in the form you want them (see Wiki). Correct me if I'm wrong, but that should mean that IRF (and FEVD) work out of the box.
Best,

**Comment by arnab13061989** [comment]
The impulse response would depend upon the matrix derived from Cholesky decomposition of Var-Covar  matrix .  The sign restrictions will be based on this matrix . Will the Cholesky decomposition be same in case of FVW transformation of residual matrix? I'm not quite sure on this

**Comment by nk027** [comment]
So in a VARX in general the IRF work the same way ([see e.g.](http://www.kurims.kyoto-u.ac.jp/EMIS/journals/RCE/V35/v35n3a09.pdf)). There the concept is the same to FWL ,so I don't see why it'd impact the VCOV matrix (and hence the Cholesky). Conceptually we're removing the correlation with `X` from all `Y`. This may impact the correlation between `y_1` and `y_2`, but this is spurious anyways and exactly why you want a VARX.

**Comment by arnab13061989** [comment]
Y = X1 *beta1 +X2* beta2 +u , X1 and X2 are independent and let X2 be the set of endogenous variables .  Now  
Mx1 *Y1 =.  Mx1*X2*beta2 +Mx1*u , where Mx1 is the annihilator matrix. Wouldn't the expressions which one will get after converting to MA representation of VAR contain the Mx1 term as coefficient of errors .......

**Comment by arnab13061989** [comment]
No my mistake , Mx1 term should not factor thanks

**Comment by arnab13061989** [comment]
yes it will help in estimation of multiplier analysis for exogenous variables . Once again thanks

**Comment by guivivi** [comment]
Hi. Thank you for the BVAR package.
Following the FWL approach, would this procedure be correct?
```
# Step 1: Regress each endogenous variable on exogenous variables
residuals_endogenous <- endogenous # Initialize residuals container with the matrix of endogenous variables
for (var in colnames(endogenous)) {
  # Linear regression: each endogenous variable ~ all exogenous variables
  model <- lm(endogenous[[var]] ~ ., data = exogenous) # exogenous is the matrix of exogenous variables
  
  # Save the residuals (endogenous variable with exogenous influence removed)
  residuals_endogenous[[var]] <- residuals(model)
}

# Step 2: Fit the BVAR model using the residuals
bvar_model <- bvar(
  residuals_endogenous, # Residuals after "partialing out" exogenous variables
  lags = 2,             # Number of lags
  n_draw = 1000,        # Posterior draws
  n_burn = 200          # Burn-in iterations
)

# Summarize and analyze the model
summary(bvar_model)
```

**Comment by guivivi** [comment]
Both the FWL and the iterative regression approaches are equivalent:

```
# Example data
set.seed(123)
exogenous <- matrix(rnorm(100), ncol = 2)  # 2 exogenous variables
endogenous <- matrix(rnorm(100), ncol = 2)  # 2 endogenous variables

# FWL Method
Q_fwl <- qr.Q(qr(exogenous))
endogenous_fwl <- endogenous - Q_fwl %*% crossprod(Q_fwl, endogenous)

# Iterative Method
residuals_endogenous <- endogenous
for (i in seq_len(ncol(endogenous))) {
  model <- lm(endogenous[, i] ~ exogenous -1)
  residuals_endogenous[, i] <- residuals(model)
}

# Compare results
all.equal(endogenous_fwl, residuals_endogenous)
```
